### ISSUE TITLE
Message is not marked as read in LHN when opening same message again

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#13740]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:

1. Open any message
2. Right click to 'Mark Unread' the message
3. Return the chat list view (message should be bolded as if it is unread)
4. Open the same chat again
5. Return the chat list view, see that chat is still bolded despite it having been read 

## Expected Result: 
Report should not be bolded in the LHN after re-opening the chat 

## Actual Result:
Report is bolded in the chat list despite having been re-opened/read

## Workaround:
unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?

- [x] iOS / native

**Version Number:** 1.2.42-0
**Reproducible in staging?:**y
**Reproducible in production?:**y
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** 

https://user-images.githubusercontent.com/43996225/208769357-54f907bf-e455-4c0c-a62e-990a0a9ae179.mov

https://user-images.githubusercontent.com/43996225/208770422-8e38a468-7150-46e0-a52e-80c744d91d81.MP4

**Expensify/Expensify Issue URL:**
**Issue reported by:** @abdulrahuman5196
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1671290524425899

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~017db106fd914a1192</li>
        <li>Upwork Job ID: 1606044643909046272</li>
        <li>Last Price Increase: 2022-12-22</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @JmillsExpensify</details>

**Comment by stephanieelliott** [comment]
Hey @kavimuru when you say "report" I think you meant `Request Money` right?

If so, I was able to repro this on iOS

**Comment by kavimuru** [comment]
@stephanieelliott Sorry, I have updated the title. I copy pasted bug from the slack as it is.

**Comment by stephanieelliott** [comment]
Cool, thanks @kavimuru! That makes sense now, updated the action steps based on that. 

I think this can be worked on externally.

**Comment by mountiny** [comment]
This is related to unread messages and sequence numbers, I am putting this on hold for once the refactor is finished, added it to the tracking list https://github.com/Expensify/Expensify/issues/227234

**Comment by Prince-Mendiratta** [comment]
Since I had already begun to test this before this was marked as out of scope, to add onto my findings to this issue anyways -

This bug can be found on all browsers as long as they have a separate LHN window. 
The bold mark doesn't go away unless either the page is refreshed and the chat is opened and then closed or another chat is opened first before opening the marked chat.

Even when the message is marked as read after opening the chat after a refresh, the --New indicator still persists on opening the app again until another chat is opened.


https://user-images.githubusercontent.com/54077356/209237907-a5c7e87b-4624-4858-989f-6778a6be295f.mp4

**Comment by onesh** [comment]
Hi team, how are we managing the state for storing the chats ?

With the information on this page I can surely say that the state does gets updated once the chat is marked as unread (that’s why when we go back, the last message is boldened) but when that chat is opened again, there should be another callback that unchecks a flag in that chat object in the state and triggers a re-render. Perhaps the best way to do that could be using useEffect if we are using hooks because that lets you mutate the state object before the component is rendered so as to prevent subsequent renders.

Could you please point me to the code/files for this bug to have a closer look?

Thanks!

**Comment by JmillsExpensify** [comment]
Added myself as the BZ member helping with bug testing/resolution once sequence numbers is wrapped up.

**Comment by abdulrahuman5196** [comment]
Hi @JmillsExpensify . Since the bug is on hold and worked by internal, what is process for bug reporter payment?. Because I haven't received upwork invite for this bug yet?

**Comment by JmillsExpensify** [comment]
Any payment would happen once the issue is taken of hold and the PR has been on production for 7 days. [More here](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#payment-for-contributions).

**Comment by abdulrahuman5196** [comment]
@JmillsExpensify  Got it. Thank you for your quick response.

**Comment by JmillsExpensify** [comment]
Still on hold for the sequence number project.

**Comment by JmillsExpensify** [comment]
This one is still outstanding. cc @chiragsalian what's our plan for the remaining unread bugs following our recent refactor?

**Comment by chiragsalian** [comment]
I think the general usecase of this is tackled. [Context](https://expensify.slack.com/archives/C036S0BVAHH/p1675376411864689?thread_ts=1674479084.856109&cid=C036S0BVAHH).
A guide pinged me about an edge case today, I asked them to post on a public channel to make it easier to discuss but they haven't yet. So I think this issue can be closed and now just specific scenarios can be tackled separately. Thoughts?

**Comment by JmillsExpensify** [comment]
Oh sorry, I was able to reproduce this yesterday on iOS using the reproduction steps above. Are you not able to for some reason?

**Comment by chiragsalian** [comment]
Ah yes, I can reproduce the issue. I didn't realize earlier that the issue was specific to marking unread.

Cool, i think this should be straightforward to solve. The solution might be as simple as checking if drawer status has changed from open to close in the bottom of this while the reportID was still the same in ReportActionsView or ReportScreen, if true then return true and call readNewestAction.

Also if this helps someone, its easily testable on web using chrome dev tools by toggling mobile view.

**Comment by MelvinBot** [comment]
Current assignee @JmillsExpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @parasharrajat is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Current assignee @amyevans is eligible for the External assigner, not assigning anyone new.

**Comment by JmillsExpensify** [comment]
Nice, thanks for circling back. I went ahead and made this one external again, hoping it gets picked up soon!

**Comment by abdulrahuman5196** [comment]
**Proposal**
**Solution**
We already check for isDrawerOpen in ReportActionsView and set the indicator status(newMarkerReportActionID) and call Report.openReport() accordingly. 
But the problem is we freeze the ReportActionsView in ReportScreen as soon as we see isDrawerOpen to be true.
https://github.com/Expensify/App/blob/26fdacee7e13cb4c9f1d563ac224658c3d65b25b/src/pages/home/ReportScreen.js#L213-L214

This means that the state of isDrawerOpen: true is not propagated to the ReportActionsView(any other props for that matter), so ReportActionsView has props of isDrawerOpen to be false. So when we open the report again and unfreeze the ReportActionsView it still has isDrawerOpen to false and the new prop is also false, so it doesn't update the unread status. 

So I think we can send the isDrawerOpen props one last time before freezing the ReportActionsView when user moves from the ReportScreen and freeze it immediately to make sure we don't break anything and cause performance issues(I think this is why freeze was introduced). This way we can ensure all props at the closing state of the ReportScreen is sent to ReportActionsView as well and can prevent any future issues similar to this.

```diff
--- a/src/pages/home/ReportScreen.js
+++ b/src/pages/home/ReportScreen.js
@@ -116,6 +116,7 @@ class ReportScreen extends React.Component {
             skeletonViewContainerHeight: reportActionsListViewHeight,
             viewportOffsetTop: 0,
             isBannerVisible: true,
+            canFreeze: true,
         };
     }
 
@@ -126,6 +127,12 @@ class ReportScreen extends React.Component {
     }
 
     componentDidUpdate(prevProps) {
+        if (this.state.canFreeze && this.props.isDrawerOpen && !prevProps.isDrawerOpen) {
+            this.setState({canFreeze: false});
+        }
+        if (!this.state.canFreeze && this.props.isDrawerOpen) {
+            this.setState({canFreeze: true});
+        }
         if (this.props.route.params.reportID === prevProps.route.params.reportID) {
             return;
         }
@@ -211,7 +218,7 @@ class ReportScreen extends React.Component {
         const isLoadingInitialReportActions = _.isEmpty(this.props.reportActions) && this.props.report.isLoadingReportActions;
 
         // When the ReportScreen is not open/in the viewport, we want to "freeze" it for performance reasons
-        const freeze = this.props.isSmallScreenWidth && this.props.isDrawerOpen;
+        const freeze = this.props.isSmallScreenWidth && this.props.isDrawerOpen && this.state.canFreeze;
 
         // the moment the ReportScreen becomes unfrozen we want to start the animation of the placeholder skeleton content
         // (which is shown, until all the actual views of the ReportScreen have been rendered)
```
We can also the above fix in shouldComponentUpdate() or render() but it seems the above is best. If we wanted the same fix at different place also i can work it out. But the core proposal is to send the props one last time to ReportActionsView before navigating back to LHN list.

**Alternative**
Another solution is to check if the isDrawerOpen status changed in ReportScreen itself and call Report.openReport() or readNewestAction in the below componentDidUpdate.
https://github.com/Expensify/App/blob/26fdacee7e13cb4c9f1d563ac224658c3d65b25b/src/pages/home/ReportScreen.js#L128-L135
But this doesn't not reset the new message marker(newMarkerReportActionID) in the ReportActionsList page as its an internal state, so we would end making changes to newMarkerReportActionID which can affect current behaviour and seems to be over-engineering for this issue.
**Alternative**
We can also totally remove the freeze component if its not useful anymore as I see we reload the total component on swipe back and forth as well.

**Comment by s77rt** [comment]
This issue has been mentioned [here](https://github.com/Expensify/App/issues/14649#issuecomment-1419339170) too. I have already provided the [root cause](https://github.com/Expensify/App/issues/14649#issuecomment-1419687291).

**TL;DR**
Hold this issue in favour of navigation reboot (it will most likely no longer be an issue)

cc @roryabraham

**Comment by MelvinBot** [comment]
Looks like something related to `react-navigation` may have been mentioned in this issue discussion.

As a reminder, please make sure that all proposals are not workarounds and that any and all attempt to fix the issue holistically have been made before proceeding with a solution. Proposals to change our `DeprecatedCustomActions.js` files should not be accepted.

Feel free to drop a note in [#expensify-open-source](https://expensify.slack.com/archives/C01GTK53T8Q) with any questions.

**Comment by amyevans** [comment]
Okay, I agree that makes sense, although @JmillsExpensify @chiragsalian @roryabraham if any of you disagree please speak up! For now adding `[HOLD React-Navigation]` to the title and moving again to `Monthly`.

**Comment by JmillsExpensify** [comment]
Well shoot, I was hoping we could address sooner, but those explanations and root cause make sense. This is one of the easier unread bugs to reproduce.

**Comment by bondydaa** [comment]
going to close this out in favor of centralizing discussion on https://github.com/Expensify/App/issues/15212 since we're going to try and solve all this more holistically with a design doc

**Comment by bondydaa** [comment]
whoops sorry I jumped the gun on closing, leaving open and on hold for now and dropping priority to monthly

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by JmillsExpensify** [comment]
Still on hold.

**Comment by amyevans** [comment]
Still on hold

**Comment by JmillsExpensify** [comment]
I need to test this, on my list now that react-navigation bugs from the first wave of implementation are starting to wind down.

**Comment by JmillsExpensify** [comment]
Ah sorry I missed your last message. That's good enough for me! I'll file this issue in those that are awaiting payment when we close out the dozen or so testing issues that still remain.

**Comment by JmillsExpensify** [comment]
Same

**Comment by JmillsExpensify** [comment]
@abdulrahuman5196 would you mind re-testing this issue, including video, and comment confirming whether it's still an issue?

**Comment by JmillsExpensify** [comment]
@abdulrahuman5196 friendly bump.

**Comment by abdulrahuman5196** [comment]
@JmillsExpensify This issue seems to be fixed.

**Comment by abdulrahuman5196** [comment]
> @JmillsExpensify This issue seems to be fixed.

Gentle Bump here

**Comment by roryabraham** [comment]
Seems like all that's left to do here is pay out the reporting bonus.

**Comment by JmillsExpensify** [comment]
Sorry for the 

> @abdulrahuman5196 would you mind re-testing this issue, including video, and comment confirming whether it's still an issue?

In the future please provide video confirming the test/fix.

**Comment by JmillsExpensify** [comment]
That being said, I will go ahead and send the Upwork offer. Thanks!

**Comment by abdulrahuman5196** [comment]
@JmillsExpensify I tested again and the issue is reproducible. It could be my bad that it was not reproducible the last time i tested.


https://github.com/Expensify/App/assets/46707890/dc8b879f-9466-419a-b965-3b26a855ce24

**Comment by JmillsExpensify** [comment]
Ok all good! So what next steps do ya'll recommend? Open this back up for proposals?

**Comment by amyevans** [comment]
Posted this [in Slack](https://expensify.slack.com/archives/C01GTK53T8Q/p1700550440543889) as well, but I cannot repro.

https://github.com/Expensify/App/assets/7277067/9ca665f6-3dfb-4b54-86e4-771764503fd2

**Comment by MonilBhavsar** [comment]
This should be fixed! We can retest and most probably close this

**Comment by JmillsExpensify** [comment]
@lanitochka17 Mind trying to reproduce this one? At that point we can close.

**Comment by JmillsExpensify** [comment]
Though actually, I think @amyevans confirmation is enough. Going to issue payment for reporting to @abdulrahuman5196.

**Comment by abdulrahuman5196** [comment]
Accepted the new offer @JmillsExpensify

**Comment by JmillsExpensify** [comment]
All paid out. Thank you!

**Comment by JmillsExpensify** [comment]
P.S. I discovered a duplicate offer, so I've just closed that one out. Sorry for the confusion.
